pkg/nodeidentity/do/identify.go

```diff
@@ -0,0 +1,166 @@
+type nodeIdentifier struct {
+	doClient *godo.Client
+}
+
+const (
+	dropletRegionMetadataURL    = "http://169.254.169.254/metadata/v1/region"
+	dropletTagInstanceGroupName = "kops-instancegroup"
+)
+
+// TokenSource implements oauth2.TokenSource
+type TokenSource struct {
+	AccessToken string
+}
+
+// Token() returns oauth2.Token
+func (t *TokenSource) Token() (*oauth2.Token, error) {
+	token := &oauth2.Token{
+		AccessToken: t.AccessToken,
+	}
+	return token, nil
+}
+
+// New creates and returns a nodeidentity.Identifier for Nodes running on OpenStack
+func New() (nodeidentity.Identifier, error) {
+	region, err := getMetadataRegion()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get droplet region: %s", err)
+	}
+
+	godoClient, err := NewCloud(region)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize digitalocean cloud: %s", err)
+	}
+
+	return &nodeIdentifier{
+		doClient: godoClient,
+	}, nil
+}
+
+func getMetadataRegion() (string, error) {
+	return getMetadata(dropletRegionMetadataURL)
+}
+
+// NewCloud returns a Cloud, expecting the env var DIGITALOCEAN_ACCESS_TOKEN
+// NewCloud will return an err if DIGITALOCEAN_ACCESS_TOKEN is not defined
+func NewCloud(region string) (*godo.Client, error) {
+	accessToken := os.Getenv("DIGITALOCEAN_ACCESS_TOKEN")
+	if accessToken == "" {
+		return nil, errors.New("DIGITALOCEAN_ACCESS_TOKEN is required")
+	}
+
+	tokenSource := &TokenSource{
+		AccessToken: accessToken,
+	}
+
+	oauthClient := oauth2.NewClient(context.TODO(), tokenSource)
+	client := godo.NewClient(oauthClient)
+
+	return client, nil
+}
+
+func getMetadata(url string) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("droplet metadata returned non-200 status code: %d", resp.StatusCode)
+	}
+
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return string(bodyBytes), nil
+}
+
+// IdentifyNode queries OpenStack for the node identity information
+func (i *nodeIdentifier) IdentifyNode(ctx context.Context, node *corev1.Node) (*nodeidentity.Info, error) {
+	providerID := node.Spec.ProviderID
+	if providerID == "" {
+		return nil, fmt.Errorf("providerID was not set for node %s", node.Name)
+	}
+	if !strings.HasPrefix(providerID, "digitalocean://") {
+		return nil, fmt.Errorf("providerID %q not recognized for node %s", providerID, node.Name)
+	}
+
+	instanceID := strings.TrimPrefix(providerID, "digitalocean://")
+	if strings.HasPrefix(instanceID, "/") {
+		instanceID = strings.TrimPrefix(instanceID, "/")
+	}
+
+	kopsGroup, err := i.getInstanceGroup(instanceID)
+	if err != nil {
+		return nil, err
+	}
+
+	info := &nodeidentity.Info{}
+	info.InstanceGroup = kopsGroup
+
+	return info, nil
+}
+
+func (i *nodeIdentifier) getInstanceGroup(instanceID string) (string, error) {
+
+	dropletID, err := strconv.Atoi(instanceID)
+	ctx := context.TODO()
```

Comment:
Does it make sense to use a timeout-based context here to make sure we do not hang too long trying to fetch the droplet?

Reply:
There are lot of places where context.TODO is used. I'll create an issue to address that as a separate PR. Please suggest if that's Ok for you.

Reply:
Totally okay with that 👍